pkg/controller/scheduler.go

```diff
@@ -322,32 +325,29 @@ func (c *Controller) advanceCanary(name string, namespace string) {
 
 	// route traffic back to primary if analysis has succeeded
 	if cd.Status.Phase == flaggerv1.CanaryPhasePromoting {
+		//start traffic shift only when the primary is ready
+		if err := canaryController.IsPrimaryReady(cd); err != nil {
+			c.recordEventWarningf(cd, "%v", err)
+			return
+		}
+
 		c.runPromotionTrafficShift(cd, canaryController, meshRouter, provider, canaryWeight, primaryWeight)
 		return
 	}
 
 	// scale canary to zero if promotion has finished
-	if cd.Status.Phase == flaggerv1.CanaryPhaseFinalising {
-		if err := canaryController.ScaleToZero(cd); err != nil {
-			c.recordEventWarningf(cd, "%v", err)
-			return
-		}
+	if (cd.Status.Phase == flaggerv1.CanaryPhaseFinalising ||
+		cd.Status.Phase == flaggerv1.CanaryPhaseWaitingFinalising) &&
+		cd.Status.FailedChecks < cd.GetAnalysisThreshold() {
 
-		// set status to succeeded
-		if err := canaryController.SetStatusPhase(cd, flaggerv1.CanaryPhaseSucceeded); err != nil {
-			c.recordEventWarningf(cd, "%v", err)
-			return
-		}
-		c.recorder.SetStatus(cd, flaggerv1.CanaryPhaseSucceeded)
-		c.runPostRolloutHooks(cd, flaggerv1.CanaryPhaseSucceeded)
-		c.recordEventInfof(cd, "Promotion completed! Scaling down %s.%s", cd.Spec.TargetRef.Name, cd.Namespace)
-		c.alert(cd, "Canary analysis completed successfully, promotion finished.",
-			false, flaggerv1.SeverityInfo)
+		c.runFinalizing(cd, canaryController)
 		return
 	}
 
-	// check if the number of failed checks reached the threshold
-	if (cd.Status.Phase == flaggerv1.CanaryPhaseProgressing || cd.Status.Phase == flaggerv1.CanaryPhaseWaitingPromotion) &&
+	// check if the number of failed checks reached the threshold for rollback
+	if (cd.Status.Phase == flaggerv1.CanaryPhaseProgressing ||
+		cd.Status.Phase == flaggerv1.CanaryPhaseWaitingPromotion ||
+		cd.Status.Phase == flaggerv1.CanaryPhaseWaitingFinalising) &&
```

Comment:
`WaitingFinalizing` comes between `Promoting` and `Finalizing`, we would not want to rollback after the canary has been promoted.